### DISCUSSION TITLE
script to auto generate files under `paths`

### DISCUSSION THREAD

**Discussion by scootcho** [dicebear/dicebear#141] [Ideas]
I was playing with this repo a few weeks ago. I was trying to use my own art to generate avatar. It was fun but took a lot of time, the most tedious part was having to manually copy, paste, configure all the SVGs into files under the `/src/paths` directory. It would be really nice if there's a way to auto-generate these files based on a file structure and provided SVG file names.

If there's a way to just drop all my SVG exports from Figma into different layers/subdirectories under something like a `src/files` directory. Where the script will recursively scan all the files and auto generate those `paths/....ts` files. The only complicated one I see is something like `top.ts` in the `/vatars-avataaars-sprites` package.

I think if we can auto generate the `paths/` files we can greatly improve the workflow of importing new avatar libraries.
```
├── src
│   ├── files
│   │   ├── accessories
│   │   |    ├── kurt.svg
│   │   |    ├── sunglasses.svg
│   │   |    ├── wayfarer.svg
│   │   |    ├── ...svg
│   │   ├── clothing
│   │   |    ├── graphicShirt.svg
│   │   |    ├── hoodie.svg
│   │   |    ├── overall.svg
│   │   |    ├── ...svg
│   │   ├── eyebrows
│   │   |    ├── ...svg
│   │   |    ├── ...svg
│   │   ├── mouth
│   │   |    ├── ...svg
│   │   |    ├── ...svg
│   │   ├── nose
│   │   |    ├── ....svg
│   ├── index.ts
│   ├── paths
│   │   ├── accessories.ts
│   │   ├── clothing.ts
│   │   ├── eyebrows.ts
│   │   ├── index.ts
│   │   ├── mouth.ts
│   │   ├── nose.ts
│   │   ├── skin.ts
│   │   └── top.t
```

`src/paths/accessories.ts`
```
import type { ValuesType } from 'utility-types';
import { Accessories } from '../options';

export const accessories: Record<ValuesType<Accessories>, (color: string) => string> = {
  kurt: (color: string) => `
    <path .../>
      ...
    <path .../>
  `,
  sunglasses: (color: string) => `
    <path .../>
  `,
  ...
```

How are you importing these SVGs into dicebear? What's the most efficient way to import a Figma project? Thanks!

**Comment by FlorianKoerner** [comment]
The process of creating custom avatar styles definitely needs to be simplified. The first step in the right direction was made this week with the release of the DiceBear CLI. This makes it very easy to create and compile new avatar style projects.

```
dicebear project new my-avatar-style
cd my-avatar-style
npm run build
```

However, the way from SVG files to TypeScript files is still quite complicated and time-consuming. This is definitely something I would like to improve.

Currently I also import SVG files via copy and paste. It's really not fun.

Your idea should be quite easy to implement. I'll take a closer look at it this weekend. My long term wish would be a Figma plugin that then also takes over colors and positions from the avatar parts. If the Figma document follows a fixed structure, no more rework in the TypeScript files would be necessary. At least in theory.

**Reply by scootcho** [reply]
Thanks for taking the time to reply and for closing the issue. I should have started a discussions thread and not under issues.

Didn't see there was a new CLI available now. I'll check it out!

It would be wonderful if you can look into the possibility of auto-generate those files under `paths/`. I saw there are some additional logic in `top.ts` for layering SVG based on z-index. But the bulk of other files are mainly copy & paste right? I spent most of the time copy and pasting SVGs and it took way longer than I had hope. It would be an awesome feature to generate those. And I assume that would greatly increase the speed of importing other libraries as well.

As for more complicated ones like `tops.ts`, I wonder if there's a way to isolate those types and use a different generating standard with some kind of z-index mapping? 

I think the Figma plugin would be awesome. I would definitely use it. It will allow designers to focus on creating art to really speed up the libraries under the DiceBear offering.

**Reply by FlorianKoerner** [reply]
A quick feedback. I have used the last few days to write a Figma plugin. I think next weekend I should be able to provide the first beta.

![Bildschirmfoto 2021-08-15 um 16 44 37](https://user-images.githubusercontent.com/647303/129482337-306dd947-215b-432b-9aee-448166043503.png)

**Reply by scootcho** [reply]
looks awesome! Looking forward to test it out!

**Reply by FlorianKoerner** [reply]
The plugin already creates an export, but unfortunately it is still buggy. I will be working on bugfixes and documentation for the next days.

The development can be seen in this repo:
https://github.com/dicebear/figma-plugins

**Reply by FlorianKoerner** [reply]
I was able to create the first completely functional export today. The documentation is still missing and the publication in the Figma community.

But you can use the plugin already today. In the attachment you will find the plugin and the Figma file I used, which explains a bit how the design must be structured.

I also add my export. Will be the next official avatar style. Ask me if you have any questions.

[plugin-and-exports.zip](https://github.com/dicebear/dicebear/files/7063023/plugin-and-exports.zip)

**Reply by FlorianKoerner** [reply]
Now in review:
https://www.figma.com/community/plugin/1005765655729342787/DiceBear-Exporter

**Reply by scootcho** [reply]
Awesome! Going to try it out this weekend!

**Reply by FlorianKoerner** [reply]
Here you can find some more compatible Figma designs:
https://github.com/dicebear/dicebear/blob/main/SOURCES.md

**Reply by FlorianKoerner** [reply]
New Plugin version:
[plugin.zip](https://github.com/dicebear/dicebear/files/7073270/plugin.zip)
